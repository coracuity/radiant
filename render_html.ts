/**
 * This module provides functions for rendering HTML elements and documents to strings.
 * @module
 */

import { BaseHTMLElement, Doctype, type HTMLDocument, VoidBaseHTMLElement } from "./html_element.ts";
import { stringifyAttributes } from "./render.ts";
import type { HTMLTag, VoidHTMLTag } from "./tags.ts";

/**
 * Renders the doctype declaration for an HTML document.
 *
 * @param _doc - The document object.
 * @returns The doctype declaration as a string.
 */
export const renderDoctype = (_doc: Doctype): string => {
  return "<!DOCTYPE html>";
};

/**
 * Renders a void HTML element as a string.
 *
 * @template T - The HTML tag type.
 * @template A - The attributes type.
 * @param {VoidBaseHTMLElement<T, A>} element - The void HTML element to render.
 * @returns {string} - The rendered HTML string.
 */
export const renderVoidElement = <T extends Extract<HTMLTag, VoidHTMLTag>, A>(
  element: VoidBaseHTMLElement<T, A>,
): string => {
  return `<${element.tag}${stringifyAttributes(element.attributes ?? {})}>`;
};

/**
 * Renders an HTML element as a string.
 *
 * @template T - The HTML tag type.
 * @template A - The attributes type.
 * @template C - The children type.
 * @param element - The HTML element to render.
 * @returns The rendered HTML element as a string.
 */
export const renderHtmlElement = <T extends HTMLTag, A, C>(
  element: BaseHTMLElement<T, A, C>,
): string => {
  let result = `<${element.tag}${stringifyAttributes(element.attributes ?? {})}>`;

  for (const child of element.children) {
    if (typeof child === "string") {
      result += child;
    } else {
      result += renderElement(child);
    }
  }

  result += `</${element.tag}>`;

  return result;
};

/**
 * Renders an HTML element or void element to a string.
 *
 * @param tag - The HTML element or void element to render.
 * @returns The rendered element as a string.
 * @throws Error if the element type is not supported.
 */
export const renderElement = (tag: unknown): string => {
  if (tag instanceof Doctype) {
    return renderDoctype(tag);
  }

  if (tag instanceof BaseHTMLElement) {
    return renderHtmlElement(tag);
  }

  if (tag instanceof VoidBaseHTMLElement) {
    return renderVoidElement(tag);
  }

  // This should never happen
  throw new Error("Unsupported element type");
};

/**
 * Renders an HTML document to a string.
 *
 * @param doc - The HTML document to render.
 * @returns The rendered HTML document as a string.
 */
export const renderDocument = (doc: HTMLDocument): string => {
  let result = "";

  for (const child of doc.children) {
    result += renderElement(child);
  }

  return result;
};
