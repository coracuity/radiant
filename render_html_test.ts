import { describe, it } from "jsr:@std/testing/bdd";
import { renderDocument } from "./render_html.ts";
import { h } from "./html_tags.ts";
import { assertSnapshot } from "jsr:@std/testing/snapshot";

describe("HTML rendering", () => {
  it("works without pretty", async (t) => {
    const input = h.document(
      h.doctype(),
      h.html(
        h.head(
          h.title("Hello, world!"),
        ),
      ),
    );

    await assertSnapshot(t, renderDocument(input));
  });

  it("works with pretty", async (t) => {
    const doc = h.document(
      h.doctype(),
      h.html(
        h.head(
          h.meta({ charset: "utf-8" }),
          h.title("Hello, world!"),
        ),
        h.body(
          h.h1({ class: "blah" }, "Hello, world!"),
          h.p("This is a bunch of text... blah blah blah"),
          h.p("With some more ", h.b("bold"), " text."),
        ),
      ),
    );

    await assertSnapshot(t, renderDocument(doc, { pretty: true }));
  });
});
