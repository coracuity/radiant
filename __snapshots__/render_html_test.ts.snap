export const snapshot = {};

snapshot[`HTML rendering > works without pretty 1`] = `"<!DOCTYPE html><html><head><title>Hello, world!</title></head></html>"`;

snapshot[`HTML rendering > works with pretty 1`] = `
'<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>Hello, world!</title>
  </head>
  <body>
    <h1 class="blah">Hello, world!</h1>
    <p>This is a bunch of text... blah blah blah</p>
    <p>
      With some more 
      <b>bold</b>
       text.
    </p>
  </body>
</html>'
`;
